### PR TITLE
Configure T5 finetuning metrics

### PR DESCRIPTION
**Pull request by MaximumEntropy**
# What does this PR do ?

Adds the ability to

1. Specify the validation metric you want to use while fine-tuning GLUE/XNLI or custom datasets
2. Write model predictions on the validation and test sets to a file even with data parallel > 0.

**Collection**: NLP

# Changelog 
- Modify setup_metric in MegatronT5FinetuneModel to configure metrics based on what the user provides in their config
- Update inference step and inference epoch ends to log with this metric accoridngly.
- Update yaml files to reflect these changes.

# Usage
model.data.validation_ds.metric_name='spearman_corr_coeff'
model.data.validation_ds.write_predictions_to_file=True
model.data.validation_ds.output_file_path_prefix='/path/to/model_predictions'

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_finetune_model.py [hunk 1]
```diff
@@ -255,97 +309,204 @@ def inference_step(self, batch, batch_idx, mode, dataloader_idx=0):
...
+        return averaged_loss, averaged_metric
+
+    def write_predictions_to_file(self, outputs, output_file_path_prefix):
+        with open(output_file_path_prefix + "_preds.txt", 'w', encoding='utf-8') as f_preds, open(
+            output_file_path_prefix + "_labels.txt", 'w', encoding='utf-8'
+        ) as f_labels, open(output_file_path_prefix + "_inputs.txt", 'w', encoding='utf-8') as f_input:
+            for pred, label, input in zip(outputs['preds'], outputs['labels'], outputs['inputs']):
+                f_preds.write(f"{pred}\n")
+                f_labels.write(f"{label}\n")
+                f_input.write(f"{input}\n")
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_finetune_model.py:497]: Sure, I can do that.

File: nemo/collections/nlp/models/language_modeling/megatron_finetune_model.py [hunk 2]
```diff
@@ -219,6 +233,46 @@ def training_step(self, batch, batch_idx):
             )
         return super().training_step(batch, batch_idx)
 
+    def cast_for_metric(self, pred, label, metric_name):
+        if metric_name == 'exact_string_match':
+            return pred, label
+        pred = pred.replace(' ', '')
+        label = label.replace(' ', '')
```
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_finetune_model.py:240]: This is a good point. I noticed in some cases the detokenization would add unnecessary spaces

[the remark this answers is not in the document]
Comment (yaoyu-33) [reply nemo/collections/nlp/models/language_modeling/megatron_finetune_model.py:240]: Cool, but i am not so sure about this. replacing multiple spaces with one space should be okay. Or its more like a space within a word?

[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_finetune_model.py:240]: Yeah this is bad. I agree. Let me try to fix it.

File: nemo/collections/nlp/models/language_modeling/megatron_finetune_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_finetune_model.py]: Fixed.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!
